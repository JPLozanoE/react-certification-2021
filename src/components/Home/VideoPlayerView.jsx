import React, { useContext } from 'react';
import {
  VideoContainer,
  ItemPlayerView,
  Videoiframe,
  Title,
  Description,
} from './styles/VideoPlayerView';
import { VideoSidebar } from './VideoSidebar';
import { AppContext } from '../../state/AppContext';
import { Button } from '../../__globalStyles';
import { types } from '../../types/types';

export const VideoPlayerView = ({ favorites }) => {
  // const [addedAsFavorite, setAddedAsFavorite] = useState(false);
  const {
    state: {
      videos,
      selectedVideo,
      favoriteVideos,
      auth: { isLogged },
    },
    dispatch,
  } = useContext(AppContext);

  const handleAddFavorite = (videoPayload) => {
    // setAddedAsFavorite((prevState) => !prevState);
    if (selectedVideo?.isFavorite) {
      dispatch({
        type: types.deleteFavoriteVideo,
        payload: videoPayload.id.videoId,
      });
      dispatch({
        type: types.deleteCurrentVideo,
        payload: videoPayload.id.videoId,
      });
    } else {
      dispatch({
        type: types.addFavoriteVideo,
        payload: videoPayload,
      });
      dispatch({
        type: types.updateRecommendedVideos,
        payload: videoPayload.id.videoId,
      });
      dispatch({
        type: types.addCurrentVideo,
      });
    }
  };

  const {
    id: { videoId },
    snippet: { title, description },
  } = selectedVideo;
  return (
    <VideoContainer>
      <ItemPlayerView>
        <Videoiframe
          width="100%"
          src={`https://www.youtube.com/embed/${videoId}`}
          frameBorder="0"
          allow="accelerometer; autoplay; clipboard-write; encrypted-media; gyroscope; picture-in-picture"
          allowFullScreen
          title="Embedded youtube"
        />
        {isLogged && (
          <Button onClick={() => handleAddFavorite(selectedVideo)} type="button">
            {selectedVideo?.isFavorite ? 'Delete from favorites' : 'Add to favorites'}
          </Button>
        )}
        <Title>{title}</Title>
        <Description>{description}</Description>
      </ItemPlayerView>
      <ItemPlayerView>
        <VideoSidebar
          videos={favorites ? favoriteVideos : videos}
          editFavorites={favorites}
        />
      </ItemPlayerView>
    </VideoContainer>
  );
};
