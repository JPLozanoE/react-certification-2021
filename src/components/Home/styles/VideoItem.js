import styled from 'styled-components';

export const VideoContainer = styled.div`
  margin: 10px;
  cursor: pointer;
`;
export const VideoTitle = styled.h3`
  color: ${(props) => props.theme.titles.color};
`;

export const VideoText = styled.p`
  color: ${(props) => props.theme.subtitles.color};
`;
