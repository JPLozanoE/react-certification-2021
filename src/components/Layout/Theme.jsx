import React, { useContext } from 'react';
import { ThemeProvider } from 'styled-components';
import { AppContext } from '../../state/AppContext';

const Theme = ({ children }) => {
  const {
    state: { isDarkTheme },
  } = useContext(AppContext);
  const theme = {
    main: {
      ...(isDarkTheme ? { backgroundColor: '#181818' } : { backgroundColor: '#ffffff' }),
    },
    titles: {
      ...(isDarkTheme ? { color: '#ffffff' } : { color: '#00000' }),
    },
    subtitles: {
      ...(isDarkTheme ? { color: 'rgb(170,170,170)' } : { color: 'rgb(96,96,96)' }),
    },
  };

  return <ThemeProvider theme={theme}>{children}</ThemeProvider>;
};

export default Theme;
