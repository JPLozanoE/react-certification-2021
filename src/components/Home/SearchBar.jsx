import React, { useContext, useState } from 'react';
import { useHistory } from 'react-router';
import { AppContext } from '../../state/AppContext';
import { types } from '../../types/types';
import { Bar, SearchButton, SearchInput } from './styles/SearchBar';

export const SearchBar = () => {
  const history = useHistory();
  const [input, setInput] = useState('');
  const { dispatch } = useContext(AppContext);

  const handleChange = (e) => {
    e.preventDefault();
    setInput(e.target.value);
  };

  const handleClick = () => {
    dispatch({
      type: types.setSelectedVideo,
      payload: null,
    });
    dispatch({
      type: types.changeSearchQuery,
      payload: input,
    });
    setInput('');
    history.push('/');
  };

  return (
    <Bar className="topnav" id="myTopnav">
      <SearchInput value={input} onChange={handleChange} placeholder="Buscar" />
      <SearchButton onClick={handleClick} name="buscar">
        buscar
      </SearchButton>
    </Bar>
  );
};
