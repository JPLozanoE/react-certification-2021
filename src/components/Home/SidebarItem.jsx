/* eslint-disable jsx-a11y/click-events-have-key-events */
/* eslint-disable jsx-a11y/no-noninteractive-element-interactions */
import React, { useContext } from 'react';
import { useHistory } from 'react-router';
import { AppContext } from '../../state/AppContext';
import { types } from '../../types/types';
import { Button } from '../../__globalStyles';
import { VideoBox, DescriptionBox } from './styles/SidebarItem';

export const SidebarItem = ({ video, isFavorite = false, editFavorites = false }) => {
  const { dispatch } = useContext(AppContext);
  const history = useHistory();

  const handleClick = () => {
    dispatch({
      type: types.setSelectedVideo,
      payload: video,
    });
    if (editFavorites) {
      history.push(`/video/favorites/${video.id.videoId}`);
    } else {
      history.push(`/video/${video.id.videoId}`);
    }
  };

  const handleAddFavorite = (videoPayload) => {
    if (isFavorite) {
      dispatch({
        type: types.deleteFavoriteVideo,
        payload: videoPayload.id.videoId,
      });
      history.push('/favorites');
    } else {
      dispatch({
        type: types.addFavoriteVideo,
        payload: videoPayload,
      });
      dispatch({
        type: types.updateRecommendedVideos,
        payload: videoPayload.id.videoId,
      });
    }
  };

  const {
    snippet: { title, thumbnails },
  } = video;
  return (
    <>
      <VideoBox>
        <img
          onClick={handleClick}
          src={thumbnails.default.url}
          alt="img"
          aria-label={video.snippet.title}
        />
        <DescriptionBox>
          {title}
          <div>
            <Button onClick={() => handleAddFavorite(video)} type="button">
              {isFavorite ? 'Delete from favorites' : 'Add to Favorites'}
            </Button>
          </div>
        </DescriptionBox>
      </VideoBox>
      <hr />
    </>
  );
};
