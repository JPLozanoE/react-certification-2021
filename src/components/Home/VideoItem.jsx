/* eslint-disable jsx-a11y/no-noninteractive-element-interactions */
/* eslint-disable jsx-a11y/click-events-have-key-events */
import React, { useContext } from 'react';
import { useHistory } from 'react-router';
import { AppContext } from '../../state/AppContext';
import { types } from '../../types/types';
import { VideoContainer, VideoTitle, VideoText } from './styles/VideoItem';
import { Button } from '../../__globalStyles';

export const VideoItem = ({ video, editFavorites, isFavorite }) => {
  const {
    dispatch,
    state: {
      auth: { isLogged },
    },
  } = useContext(AppContext);
  const history = useHistory();

  const handleVideoClick = (videoPayload) => {
    dispatch({
      type: types.setSelectedVideo,
      payload: videoPayload,
    });
    if (editFavorites) {
      history.push(`/video/favorites/${video.id.videoId}`);
    } else {
      history.push(`/video/${video.id.videoId}`);
    }
  };

  const handleAddFavorite = (videoPayload) => {
    if (editFavorites) {
      dispatch({
        type: types.deleteFavoriteVideo,
        payload: videoPayload.id.videoId,
      });
    } else {
      dispatch({
        type: types.addFavoriteVideo,
        payload: videoPayload,
      });
      dispatch({
        type: types.updateRecommendedVideos,
        payload: videoPayload.id.videoId,
      });
    }
  };

  return (
    <VideoContainer>
      <img
        width="100%"
        height="150px"
        style={{ objectFit: 'cover' }}
        src={video.snippet.thumbnails.high.url}
        alt="img"
        aria-label={video.snippet.title}
        onClick={() => handleVideoClick(video)}
      />
      <div style={{ padding: '10px' }}>
        <div>
          <VideoTitle onClick={() => handleVideoClick(video)}>
            {video.snippet.title}
          </VideoTitle>
          {isLogged && (
            <Button
              disabled={isFavorite && !editFavorites}
              onClick={() => handleAddFavorite(video)}
              type="button"
            >
              {editFavorites ? 'Delete from favorites' : 'Add to favorites'}
            </Button>
          )}
        </div>
        <VideoText>{video.snippet.channelTitle}</VideoText>
        <VideoText>Publicado el {video.snippet.publishTime}</VideoText>
      </div>
    </VideoContainer>
  );
};
