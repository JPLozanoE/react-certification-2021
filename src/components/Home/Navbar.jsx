import React, { useContext } from 'react';
import { SearchBar } from './SearchBar';
import { AppContext } from '../../state/AppContext';
import { types } from '../../types/types';
import {
  Bar,
  Links,
  CheckBoxWrapper,
  Text,
  CheckBoxLabel,
  CheckBox,
  DivDarkMode,
  DivSwitch,
} from './styles/NavBar';

export const Navbar = ({ setSelectedVideo }) => {
  const {
    state: {
      darkTheme,
      auth: { isLogged },
    },
    dispatch,
  } = useContext(AppContext);

  const handleClick = () => {
    dispatch({ type: types.setSelectedVideo, payload: null });
  };

  const handleLogout = () => {
    dispatch({ type: types.logout });
  };

  const handleToggleChange = () => {
    dispatch({
      type: types.toggleTheme,
    });
  };

  return (
    <Bar className="topnav" id="myTopnav">
      <Links onClick={handleClick} to="/">
        Inicio
      </Links>
      {/* <Links href="#" className="active" /> */}
      <SearchBar setSelectedVideo={setSelectedVideo} />
      <DivDarkMode>
        <Text href="#about">Oscuro</Text>
        <DivSwitch>
          <CheckBoxWrapper>
            <CheckBox
              defaultChecked={darkTheme}
              id="checkbox"
              type="checkbox"
              onChange={handleToggleChange}
            />
            <CheckBoxLabel htmlFor="checkbox" />
          </CheckBoxWrapper>
          {isLogged ? (
            <>
              <Links onClick={handleLogout} to="/">
                Logout
              </Links>
              <Links to="/favorites">Favorites</Links>
            </>
          ) : (
            <Links to="/auth">Login</Links>
          )}
        </DivSwitch>
      </DivDarkMode>
    </Bar>
  );
};
