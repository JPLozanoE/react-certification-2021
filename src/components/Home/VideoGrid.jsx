import React from 'react';
import styled from 'styled-components';
import { VideoItem } from './VideoItem';

const VideoContainer = styled.div`
  display: grid;
  grid-template-columns: auto auto auto auto;

  @media only screen and (max-width: 1200px) {
    grid-template-columns: auto auto auto;
  }

  @media only screen and (max-width: 768px) {
    grid-template-columns: auto auto;
  }

  @media only screen and (max-width: 600px) {
    grid-template-columns: auto;
  }
`;

export const VideoGrid = ({ items, setSelectedVideo }) => {
  return (
    <VideoContainer>
      {items.map((video) => (
        <VideoItem setSelectedVideo={setSelectedVideo} {...video} key={video.etag} />
      ))}
    </VideoContainer>
  );
};
