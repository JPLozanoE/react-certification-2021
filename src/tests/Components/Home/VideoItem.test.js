/* eslint-disable react/jsx-filename-extension */
import React from 'react';
import { render } from '@testing-library/react';
import { BrowserRouter as Router } from 'react-router-dom';
import { AppContext } from '../../../state/AppContext';
import Theme from '../../../components/Layout/Theme';
import { demoState, singleVideo } from '../../fixtures/demoState';
import { VideoItem } from '../../../components/Home/VideoItem';

describe('VideoItem not being logged in', () => {
  let wrapper;
  beforeEach(() => {
    const providerValues = {
      state: { ...demoState },
      dispatch: jest.fn(),
    };
    wrapper = render(
      <AppContext.Provider value={providerValues}>
        <Theme>
          <Router>
            <VideoItem video={singleVideo} editFavorites={false} isFavorite={false} />
          </Router>
        </Theme>
      </AppContext.Provider>
    );
  });

  test('Should render properly', () => {
    expect(wrapper).toMatchSnapshot();
  });
  test('Should contain a channel title', () => {
    const ChannelTitleElement = wrapper.getByText(singleVideo.snippet.channelTitle);
    expect(ChannelTitleElement).toBeInTheDocument();
  });

  test('Should contain a title', () => {
    const TitleElement = wrapper.getByText(singleVideo.snippet.title);
    expect(TitleElement).toBeInTheDocument();
  });

  test('Should contain an Image', () => {
    const ImageElement = wrapper.getByRole('img', { name: singleVideo.snippet.title });
    expect(ImageElement).toBeInTheDocument();
  });
});

describe('VideoItem being logged in', () => {
  let wrapper;
  beforeEach(() => {
    const providerValues = {
      state: { ...demoState, auth: { ...demoState.auth, isLogged: true } },
      dispatch: jest.fn(),
    };
    wrapper = render(
      <AppContext.Provider value={providerValues}>
        <Theme>
          <Router>
            <VideoItem video={singleVideo} editFavorites={false} isFavorite={false} />
          </Router>
        </Theme>
      </AppContext.Provider>
    );
  });

  test('Should contain an "Add to favorites" button ', () => {
    const ChannelTitleElement = wrapper.getByText('Add to favorites');
    expect(ChannelTitleElement).toBeInTheDocument();
  });
});

describe('VideoItem when present on favorite videos screen', () => {
  let wrapper;
  beforeEach(() => {
    const providerValues = {
      state: { ...demoState, auth: { ...demoState.auth, isLogged: true } },
      dispatch: jest.fn(),
    };
    wrapper = render(
      <AppContext.Provider value={providerValues}>
        <Theme>
          <Router>
            <VideoItem video={singleVideo} editFavorites isFavorite />
          </Router>
        </Theme>
      </AppContext.Provider>
    );
  });

  test('Should contain an "Delete favorites" button ', () => {
    const ChannelTitleElement = wrapper.getByText('Delete from favorites');
    expect(ChannelTitleElement).toBeInTheDocument();
  });
});
