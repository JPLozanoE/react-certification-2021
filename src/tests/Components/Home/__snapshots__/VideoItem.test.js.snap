// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`VideoItem not being logged in Should render properly 1`] = `
Object {
  "asFragment": [Function],
  "baseElement": <body>
    <div>
      <div
        class="sc-bdnxRM jEaRYL"
      >
        <img
          alt="img"
          aria-label="¿Qué es React JS? Breve explicación animada"
          height="150px"
          src="https://i.ytimg.com/vi/1-2eyS0k-IA/hqdefault.jpg"
          style="object-fit: cover;"
          width="100%"
        />
        <div
          style="padding: 10px;"
        >
          <div>
            <h3
              class="sc-gtsrHT gtXmqm"
            >
              ¿Qué es React JS? Breve explicación animada
            </h3>
          </div>
          <p
            class="sc-dlnjwi OrDgl"
          >
            Bitech Studio
          </p>
          <p
            class="sc-dlnjwi OrDgl"
          >
            Publicado el 
            2018-02-20T04:15:30Z
          </p>
        </div>
      </div>
    </div>
  </body>,
  "container": <div>
    <div
      class="sc-bdnxRM jEaRYL"
    >
      <img
        alt="img"
        aria-label="¿Qué es React JS? Breve explicación animada"
        height="150px"
        src="https://i.ytimg.com/vi/1-2eyS0k-IA/hqdefault.jpg"
        style="object-fit: cover;"
        width="100%"
      />
      <div
        style="padding: 10px;"
      >
        <div>
          <h3
            class="sc-gtsrHT gtXmqm"
          >
            ¿Qué es React JS? Breve explicación animada
          </h3>
        </div>
        <p
          class="sc-dlnjwi OrDgl"
        >
          Bitech Studio
        </p>
        <p
          class="sc-dlnjwi OrDgl"
        >
          Publicado el 
          2018-02-20T04:15:30Z
        </p>
      </div>
    </div>
  </div>,
  "debug": [Function],
  "findAllByAltText": [Function],
  "findAllByDisplayValue": [Function],
  "findAllByLabelText": [Function],
  "findAllByPlaceholderText": [Function],
  "findAllByRole": [Function],
  "findAllByTestId": [Function],
  "findAllByText": [Function],
  "findAllByTitle": [Function],
  "findByAltText": [Function],
  "findByDisplayValue": [Function],
  "findByLabelText": [Function],
  "findByPlaceholderText": [Function],
  "findByRole": [Function],
  "findByTestId": [Function],
  "findByText": [Function],
  "findByTitle": [Function],
  "getAllByAltText": [Function],
  "getAllByDisplayValue": [Function],
  "getAllByLabelText": [Function],
  "getAllByPlaceholderText": [Function],
  "getAllByRole": [Function],
  "getAllByTestId": [Function],
  "getAllByText": [Function],
  "getAllByTitle": [Function],
  "getByAltText": [Function],
  "getByDisplayValue": [Function],
  "getByLabelText": [Function],
  "getByPlaceholderText": [Function],
  "getByRole": [Function],
  "getByTestId": [Function],
  "getByText": [Function],
  "getByTitle": [Function],
  "queryAllByAltText": [Function],
  "queryAllByDisplayValue": [Function],
  "queryAllByLabelText": [Function],
  "queryAllByPlaceholderText": [Function],
  "queryAllByRole": [Function],
  "queryAllByTestId": [Function],
  "queryAllByText": [Function],
  "queryAllByTitle": [Function],
  "queryByAltText": [Function],
  "queryByDisplayValue": [Function],
  "queryByLabelText": [Function],
  "queryByPlaceholderText": [Function],
  "queryByRole": [Function],
  "queryByTestId": [Function],
  "queryByText": [Function],
  "queryByTitle": [Function],
  "rerender": [Function],
  "unmount": [Function],
}
`;
